File: fixtures/scale/dart/large/lib/ciphers/railfence/railfence_cipher.dart
import '../../core/cipher.dart';
import '../../util/alphabet.dart';
import 'rail_pattern.dart';
import 'railfence_key.dart';

/// Writes the text in a zigzag across rails, then reads rail by rail.
class RailFenceCipher implements Cipher {
  final RailPattern _pattern;

  RailFenceCipher(RailFenceKey key) : _pattern = RailPattern(key.rails);

  @override
  String name() => 'railfence';

  @override
  String encrypt(String plaintext) {
    final text = cleanAlphabet(plaintext);
    final rows = List.generate(_pattern.railCount(), (_) => StringBuffer());
    for (var i = 0; i < text.length; i++) {
      rows[_pattern.railFor(i)].write(text[i]);
    }
    return rows.map((r) => r.toString()).join();
  }

  @override
  String decrypt(String ciphertext) {
    final text = cleanAlphabet(ciphertext);
    final out = List<String>.filled(text.length, '');
    var cursor = 0;
    for (var r = 0; r < _pattern.railCount(); r++) {
      for (var i = 0; i < text.length; i++) {
        if (_pattern.railFor(i) == r) out[i] = text[cursor++];
      }
    }
    return out.join();
  }
}

File: fixtures/scale/java/large/src/cryptobench/ciphers/railfence/RailFenceCipher.java
package cryptobench.ciphers.railfence;

import cryptobench.core.Cipher;
import cryptobench.util.Alphabet;

/** Writes the text in a zigzag across rails, then reads rail by rail. */
public class RailFenceCipher implements Cipher {
    private final RailPattern pattern;

    public RailFenceCipher(RailFenceKey key) { this.pattern = new RailPattern(key.getRails()); }

    @Override
    public String name() { return "railfence"; }

    @Override
    public String encrypt(String plaintext) {
        String text = Alphabet.clean(plaintext);
        StringBuilder[] rows = new StringBuilder[pattern.railCount()];
        for (int r = 0; r < rows.length; r++) rows[r] = new StringBuilder();
        for (int i = 0; i < text.length(); i++) {
            rows[pattern.railFor(i)].append(text.charAt(i));
        }
        StringBuilder sb = new StringBuilder();
        for (StringBuilder row : rows) sb.append(row);
        return sb.toString();
    }

    @Override
    public String decrypt(String ciphertext) {
        String text = Alphabet.clean(ciphertext);
        char[] out = new char[text.length()];
        int cursor = 0;
        for (int r = 0; r < pattern.railCount(); r++) {
            for (int i = 0; i < text.length(); i++) {
                if (pattern.railFor(i) == r) out[i] = text.charAt(cursor++);
            }
        }
        return new String(out);
    }
}
